### DISCUSSION TITLE
Tls 1.2

### DISCUSSION THREAD

**Discussion by gpauk** [wqweto/VbAsyncSocket#20] [Q&A]
Hi..
I have an app  that works well, based from the barebonetls server example. 

It works fine with Tls1.3, except that I now have a client that can at best manage tls 1.2

If I connect from openssl with:

openssl s_client -no_tls1_3 -tlsextdebug 192.168.1.107:10443

I get a handshake error "Missing supported ciphersuite"

I found 2 instances where the code seemed to want tls1.3 only,:

`.LocalFeatures = ucsTlsSupportTls13
      
  For Each vElem In pvTlsGetSortedCipherSuites(ucsTlsSupportTls13)
      cCipherPrefs.Add cCipherPrefs.Count, "#" & vElem
  Next
`
in mdTlsThunks module.

I altered those to "ucsTlsSupportAll"

I now get the error:

"Received fatal alert. Unsupported expension"

So I may or may not be on the right track!

Any advice on how to get it to work with openssl tls1.2 would be greatly appreciated!

Thanks!

**Comment by wqweto** [comment]
The typo in the error message is hilarious :)) and I'll fix it ASAP

Unfortunately thunks based backend does not support TLS 1.2 in server mode (only client)

It's only native OS based one that supports older protocol versions but lacks TLS 1.3 support on anything but latest Win10 releases (not tested)

So currently there is not much that can be done short of impl TLS 1.2 server mode in thunks backend which is not trivial.

**Reply by gpauk** [reply]
Hi...
At least the typo is obvious - and has the advantage that makes the message easy to search for!
Thanks for the info - at least that explains why I wasn't getting anywhere...

Cheers...

**Reply by wqweto** [reply]
I just pushed commit aebec3e4b935029ffd0e992ad6956cb8af4c8032 to `master` branch which adds initial support for TLS 1.2 in server-mode.

The implementation is not very much tested but seems to work with `Secure` test project and

    c:> openssl s_client -no_tls1_3 -tlsextdebug -debug localhost:10443

. . . from command line here.

Feedback is welcome!

**Reply by gpauk** [reply]
Great! Thanks - I'll give it a try...

**Reply by wqweto** [reply]
Needed more fixes for the Barebone self-signed certificate to work w/ TLS 1.2 server-mode.

Should work now since commit eed73562ecf1eecf1885b4cd6c2b58a40737cd16

**Reply by gpauk** [reply]
Hi...
Thanks! 
Seems to be working now - you're a star!

**Reply by gpauk** [reply]
Hi...
With a client using java 8 (it's a raspberry pi zero and won't run later java) I'm getting either:
Critical error(5): Received fatal alert. Handshake failure. in TlsHandshake 
Critical error(1): Received fatal alert. Illegal parameter. in TlsHandshake 

Whereas openssl accepts the connection (see below)/
Is there a way I can refine these errors to something specific -so I can tweak client or server to get it to work?

Thanks for any suggestions!

`C:\Temp\PortableGit\mingw64\bin>openssl s_server -cert ..\..\usr\ssl\cert.pem -key ..\..\usr\ssl\key.pem
Using default temp DH parameters
ACCEPT
-----BEGIN SSL SESSION PARAMETERS-----
MHoCAQECAgMDBALAMAQgoHixBTlXQHaN5s8w9+Fw7xJpGywZAsA1X3WLyLDMPt4E
MJwIeRXVvRtDuFv46nMF1/HwOdLp0RrWeHSvB33tnIX+PBD5FCsYC4wuXz0lwsj5
aKEGAgRg4uszogQCAhwgpAYEBAEAAACtAwIBAQ==
-----END SSL SESSION PARAMETERS-----
Shared ciphers:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:AES256-SHA256:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:AES128-SHA
Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:DSA+SHA256:ECDSA+SHA1:RSA+SHA1:DSA+SHA1
Shared Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:DSA+SHA256:ECDSA+SHA1:RSA+SHA1:DSA+SHA1
Supported Elliptic Curve Point Formats: uncompressed
Supported Elliptic Groups: P-256:P-384:P-521:0x0100:0x0101:0x0102:0x0103:0x0104
Shared Elliptic groups: P-256:P-384:P-521
CIPHER is ECDHE-RSA-AES256-GCM-SHA384
Secure Renegotiation IS supported`

**Reply by wqweto** [reply]
It says the session is using `TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384` cipher which is supported by mdTlsThunks. The error says "received fatal alert" so the alert is generated by the client.

I'll have to test the server-side implementation with more clients here like Firefox, etc. when time permits, if the error can be emulated with available clients. Not sure if I'll be able to test a java client though.

**Reply by gpauk** [reply]
Thanks... I'll see if I can create the same error in a way that can be reproduced easily

**Reply by wqweto** [reply]
I think I found the issue: there was a wrong session ID used in Server Hello message for TLS 1.2 which is fixed in 3ac82f85cad901b7e98aec0b9fd012a7e4471dbf

Just tested TLS 1.2 server-mode with Chrome, Firefox, openssl and SChannel (via Windows built-in curl.exe) all seems to work ok.

**Reply by gpauk** [reply]
Thanks!
Well that's changed what I see - 

Critical error(4): Received fatal alert. Unexpected message. in TlsHandshake

I'll try and get a test program together....

**Reply by wqweto** [reply]
Just tested `Project1` in `test/BareboneTls` with Chome and Firefox and it works after accepting the self-signed certificate.

Had to modify `Form2` to call `TlsInitServer` with `LocalFeatures:=ucsTlsSupportTls12` to be sure the server-mode TLS is using version 1.2 of the protocol.

**Reply by gpauk** [reply]
Yes - that works for me too - but still fails with java 8 client.
Something about java 8 ssl support is causing a problem I'm trying to find a non java way to reproduce it - you don't want to have to install java 8!! 
Unless there's a way I can initiate a connection to you?
Thanks for trying!

**Reply by wqweto** [reply]
I found [this sample gist](https://gist.github.com/michalbcz/4170520) doing an https request in java while ignoring certificate errors.

Unfortunately I cannot execute it in one of those online compilers because of different reasons (mostly security).

**Reply by gpauk** [reply]
No worries - I thought it was unlikely - but worth suggesting!
That sample code is pretty similar. It works fine in j11 - but not in j8 which is frustrating!
I've tried forcing SSLContext.getInstance("TLS");  to TLS 1.2 and some other variants to no avail. 
I'll keep poking at it... Thanks again for your help!

**Reply by wqweto** [reply]
Managed to install JDK on a VM here
```
C:\Work\Java>java --version
java 16.0.1 2021-04-20
Java(TM) SE Runtime Environment (build 16.0.1+9-24)
Java HotSpot(TM) 64-Bit Server VM (build 16.0.1+9-24, mixed mode, sharing)
```
. . . and just tested the sample gist above with the barebone TLS server

```
C:\Work\Java>java SSLTest
200
```
. . . and it seems to work ok.

Also tested the java client with both TLS 1.3 and TLS 1.2 servers with a self-signed RSA certificate to check out negotiated ciphersuites and signatures
```
TLS 1.3:
119431.254 [INFO] Listening on 0.0.0.0:10443 [Form1.Form_Load]
119439.258 [INFO] With exchange group X25519 [mdTlsThunks.pvTlsParseHandshakeClientHello]
119439.260 [INFO] Using TLS_AES_128_GCM_SHA256 from 172.17.17.155 [mdTlsThunks.pvTlsParseHandshakeClientHello]
119439.269 [INFO] Signing with RSA_PSS_RSAE_SHA256 signature [mdTlsThunks.pvTlsSignatureSign]
119439.371 [INFO] Path=/ [cRequestHandler.HandleRequest]
119439.429 [INFO] User canceled (TLS_ALERT_LEVEL_WARNING) [mdTlsThunks.pvTlsParseRecord]
119439.431 [INFO] Close notify (TLS_ALERT_LEVEL_WARNING) [mdTlsThunks.pvTlsParseRecord]
```
```
TLS 1.2:
119476.038 [INFO] Listening on 0.0.0.0:10443 [Form1.Form_Load]
119483.768 [INFO] Using ECDHE-RSA-AES128-GCM-SHA256 from 172.17.17.155 [mdTlsThunks.pvTlsParseHandshakeClientHello]
119483.773 [INFO] Signing with RSA_PSS_RSAE_SHA256 signature [mdTlsThunks.pvTlsSignatureSign]
119484.087 [INFO] Path=/ [cRequestHandler.HandleRequest]
119484.112 [INFO] Close notify (TLS_ALERT_LEVEL_WARNING) [mdTlsThunks.pvTlsParseRecord]
```
. . . and both seem to work too.

**Reply by gpauk** [reply]
Yes - recent java is fine. 
The client in question though is on a raspberry pi zero - and java 8 is the most recent one that is supported on the arm 6 cpu on that.

It's an IoT application - needs a small low power cpu.

However - your modified internal version (samples\tlssockettest), not the barebone tls version, does connect ok!

It works rather differently from the thunks version - I just need to sort the way to send data from server to client indexed by Client.ID rather than the index in the barebone tls... 

Thanks again!

**Reply by wqweto** [reply]
I see what's going on. The TlsSocketTest sample is using different TLS backend: `mdTlsNative`.

This backend is using OS provided SChannel library for TLS implementation so it supports older versions of the protocol (TLS 1.0 and TLS 1.1) too.

All TLS backend modules (`mdTlsNative`, `mdTlsThunks` and `mdTlsSodium`) are source compatible so that these can be easily swapped if needed but differ in supported protocol features. Also `mdTlsNative` depends on Windows version so no TLS 1.2 on XP for instance.

Btw, there is `Project2.vbp` project files in the BareboneTls folder which references `mdTlsNative` already for testing puposes. `Project3.vbp` references `mdTlsSodium` while `Project4.vbp` uses a "stub protocol" (empty implementation) just to be able to measure backend modules footprint on final executable by comparing all projects compile output size to the "stub protocol" minimal one.

**Reply by gpauk** [reply]
Thanks for that - I'll have a longer look at the backend modules. Having got mdtlsthunks running (before having to support the rpi client) I hadn't really looked properly at the others. I'll spend some more time on it!

**Reply by gpauk** [reply]
Thanks - yes, project2 is a great help! Apologies for missing it before!

**Reply by wqweto** [reply]
> Unfortunately thunks based backend does not support TLS 1.2 in server mode (only client)
> 
> It's only native OS based one that supports older protocol versions but lacks TLS 1.3 support on anything but latest Win10 releases (not tested)
> 
> So currently there is not much that can be done short of impl TLS 1.2 server mode in thunks backend which is not trivial.

JFYI, I found some time (bit the bullet) and recent VB6+thunks backend supports client and **server** mode for both TLS 1.3 **and (legacy) TLS 1.2** incl. in the Barebones sample.

**Comment by gpauk** [comment]
Hi...

Great - thanks!

Graham



____________________________________________
